databus/src/main/java/com/bazaarvoice/emodb/databus/core/DefaultFanout.java

```diff
@@ -128,72 +130,80 @@ protected void shutDown() throws Exception {
-            } catch (UnknownTableException e) {
-                continue;
-            }
+                SubscriptionEvaluator.MatchEventData matchEventData;
+                try {
+                    matchEventData = _subscriptionEvaluator.getMatchEventData(eventData);
+                } catch (UnknownTableException e) {
+                    continue;
+                }
 
-            // Copy to subscriptions in the current data center.
-            for (OwnedSubscription subscription : _subscriptionEvaluator.matches(subscriptions, matchEventData)) {
-                eventsByChannel.put(subscription.getName(), eventData);
-            }
+                // Copy to subscriptions in the current data center.
+                for (OwnedSubscription subscription : _subscriptionEvaluator.matches(subscriptions, matchEventData)) {
+                    eventsByChannel.put(subscription.getName(), eventData);
+                }
 
-            // Copy to queues for eventual delivery to remote data centers.
-            if (_replicateOutbound) {
-                for (DataCenter dataCenter : matchEventData.getTable().getDataCenters()) {
-                    if (!dataCenter.equals(_currentDataCenter)) {
-                        String channel = ChannelNames.getReplicationFanoutChannel(dataCenter);
-                        eventsByChannel.put(channel, eventData);
-                        numOutboundReplicationEvents++;
+                // Copy to queues for eventual delivery to remote data centers.
+                if (_replicateOutbound) {
+                    for (DataCenter dataCenter : matchEventData.getTable().getDataCenters()) {
+                        if (!dataCenter.equals(_currentDataCenter)) {
+                            String channel = ChannelNames.getReplicationFanoutChannel(dataCenter);
+                            eventsByChannel.put(channel, eventData);
+                            numOutboundReplicationEvents++;
+                        }
                     }
                 }
-            }
 
-            // Flush to cap the amount of memory used to buffer events.
-            if (eventsByChannel.size() >= FLUSH_EVENTS_THRESHOLD) {
-                flush(eventKeys, eventsByChannel, numOutboundReplicationEvents);
-                numOutboundReplicationEvents = 0;
+                // Flush to cap the amount of memory used to buffer events.
+                if (eventsByChannel.size() >= FLUSH_EVENTS_THRESHOLD) {
+                    flush(eventKeys, eventsByChannel, numOutboundReplicationEvents);
+                    numOutboundReplicationEvents = 0;
+                }
+
+                // Track the final match event data record returned
+                lastMatchEventData = matchEventData;
             }
 
-            // Track the final match event data record returned
-            lastMatchEventData = matchEventData;
+            // Final flush.
+            flush(eventKeys, eventsByChannel, numOutboundReplicationEvents);
```

Comment:
Within `flush()` there's a call to `_eventSource.delete(eventKeys)`.  Might be good to have a timer around that as well.